File: py/configs/forwarder_config.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from configs import forwarder_config_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class NtripClient:
    casterHost: str = ""
    casterPort: int = 0
    mountPoint: str = ""
    username: str = ""
    password: str = ""
    
    _proto_message: forwarder_config_pb2.NtripClient = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = forwarder_config_pb2.NtripClient()

    def _populate_proto(self):
        self._proto_message.casterHost = self.casterHost
        self._proto_message.casterPort = self.casterPort
        self._proto_message.mountPoint = self.mountPoint
        self._proto_message.username = self.username
        self._proto_message.password = self.password

    @classmethod
    def _from_proto(cls, proto_message) -> "NtripClient":
        return cls(
            casterHost = proto_message.casterHost,
            casterPort = proto_message.casterPort,
            mountPoint = proto_message.mountPoint,
            username = proto_message.username,
            password = proto_message.password,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "NtripClient":
        message = forwarder_config_pb2.NtripClient()
        message.ParseFromString(data)
        return cls(
            casterHost = message.casterHost,
            casterPort = message.casterPort,
            mountPoint = message.mountPoint,
            username = message.username,
            password = message.password,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "NtripClient":
        message = forwarder_config_pb2.NtripClient()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class GpsDev:
    address: str = ""
    port: str = ""
    mode: str = ""
    enabled: bool = False
    
    _proto_message: forwarder_config_pb2.GpsDev = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = forwarder_config_pb2.GpsDev()

    def _populate_proto(self):
        self._proto_message.address = self.address
        self._proto_message.port = self.port
        self._proto_message.mode = self.mode
        self._proto_message.enabled = self.enabled

    @classmethod
    def _from_proto(cls, proto_message) -> "GpsDev":
        return cls(
            address = proto_message.address,
            port = proto_message.port,
            mode = proto_message.mode,
            enabled = proto_message.enabled,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "GpsDev":
        message = forwarder_config_pb2.GpsDev()
        message.ParseFromString(data)
        return cls(
            address = message.address,
            port = message.port,
            mode = message.mode,
            enabled = message.enabled,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "GpsDev":
        message = forwarder_config_pb2.GpsDev()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class ForwarderConfig:
    gpsDevice: GpsDev = None
    ntripClient: NtripClient = None
    
    _proto_message: forwarder_config_pb2.ForwarderConfig = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = forwarder_config_pb2.ForwarderConfig()

    def _populate_proto(self):
        if self.gpsDevice:
            self.gpsDevice._populate_proto()
            self._proto_message.gpsDevice.CopyFrom(self.gpsDevice._proto_message)
        if self.ntripClient:
            self.ntripClient._populate_proto()
            self._proto_message.ntripClient.CopyFrom(self.ntripClient._proto_message)

    @classmethod
    def _from_proto(cls, proto_message) -> "ForwarderConfig":
        return cls(
            gpsDevice = GpsDev._from_proto(proto_message.gpsDevice),
            ntripClient = NtripClient._from_proto(proto_message.ntripClient),
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "ForwarderConfig":
        message = forwarder_config_pb2.ForwarderConfig()
        message.ParseFromString(data)
        return cls(
            gpsDevice = (
                GpsDev._from_proto(message.gpsDevice)
                if message.HasField("gpsDevice")
                else None
            ),
            ntripClient = (
                NtripClient._from_proto(message.ntripClient)
                if message.HasField("ntripClient")
                else None
            ),
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "ForwarderConfig":
        message = forwarder_config_pb2.ForwarderConfig()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

File: src/configs/forwarder_config.cpp
#include "configs/forwarder_config.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace Configs
{
NtripClient::NtripClient(const PbConfigs::NtripClient& protobuf) {
    casterHost = protobuf.casterhost();
    casterPort = protobuf.casterport();
    mountPoint = protobuf.mountpoint();
    username = protobuf.username();
    password = protobuf.password();
}

NtripClient::operator PbConfigs::NtripClient() const {
    PbConfigs::NtripClient ret;
    ret.set_casterhost(casterHost);
    ret.set_casterport(casterPort);
    ret.set_mountpoint(mountPoint);
    ret.set_username(username);
    ret.set_password(password);
    return ret;
}

std::string NtripClient::serializeAsJsonString() const {
    PbConfigs::NtripClient protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string NtripClient::serializeAsProtobufString() const {
    PbConfigs::NtripClient protobuf(*this);
    return protobuf.SerializeAsString();
}

bool NtripClient::deserializeFromJsonString(const std::string& str) {
    PbConfigs::NtripClient protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool NtripClient::deserializeFromProtobufString(const std::string& str) {
    PbConfigs::NtripClient protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

GpsDev::GpsDev(const PbConfigs::GpsDev& protobuf) {
    address = protobuf.address();
    port = protobuf.port();
    mode = protobuf.mode();
    enabled = protobuf.enabled();
}

GpsDev::operator PbConfigs::GpsDev() const {
    PbConfigs::GpsDev ret;
    ret.set_address(address);
    ret.set_port(port);
    ret.set_mode(mode);
    ret.set_enabled(enabled);
    return ret;
}

std::string GpsDev::serializeAsJsonString() const {
    PbConfigs::GpsDev protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string GpsDev::serializeAsProtobufString() const {
    PbConfigs::GpsDev protobuf(*this);
    return protobuf.SerializeAsString();
}

bool GpsDev::deserializeFromJsonString(const std::string& str) {
    PbConfigs::GpsDev protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool GpsDev::deserializeFromProtobufString(const std::string& str) {
    PbConfigs::GpsDev protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

ForwarderConfig::ForwarderConfig(const PbConfigs::ForwarderConfig& protobuf) {
    gpsDevice = protobuf.gpsdevice();
    ntripClient = protobuf.ntripclient();
}

ForwarderConfig::operator PbConfigs::ForwarderConfig() const {
    PbConfigs::ForwarderConfig ret;
    *(ret.mutable_gpsdevice()) = gpsDevice;
    *(ret.mutable_ntripclient()) = ntripClient;
    return ret;
}

std::string ForwarderConfig::serializeAsJsonString() const {
    PbConfigs::ForwarderConfig protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string ForwarderConfig::serializeAsProtobufString() const {
    PbConfigs::ForwarderConfig protobuf(*this);
    return protobuf.SerializeAsString();
}

bool ForwarderConfig::deserializeFromJsonString(const std::string& str) {
    PbConfigs::ForwarderConfig protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool ForwarderConfig::deserializeFromProtobufString(const std::string& str) {
    PbConfigs::ForwarderConfig protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}